### PR TITLE
[alphonse] Round-3: area-weighted surface MSE loss (phys-motivated)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Each surface point has an associated area (`surface_area.npy`) that represents the physical patch of car surface it covers. Currently the surface MSE loss treats all surface points equally — a tiny near-stagnation-point patch contributes the same gradient as a large flat-panel patch. This is physically wrong: large-area patches integrate more total force on the vehicle, so getting them right matters more for drag/lift prediction.

**Hypothesis:** Weighting each surface point's loss contribution by its normalized area will steer the gradient toward physically important surface regions, improving `surface_pressure_rel_l2_pct` (and potentially all surface axes) without any architectural change.

**Why this hasn't been tried:** `surface_area` is already available in the input features (`surface_x` channel 6 = area). The loader reads `surface_area.npy` and includes it. We just need to weight the MSE by it.

**Expected effect:** Surface pressure is currently 7.64 vs AB-UPT target 3.82 (2.0× gap) — the largest remaining single-axis gap in absolute terms. Area weighting should reduce the contribution of isolated small patches (near stagnation point, trailing edge tips) that are hard to predict but physically less important.

## Instructions

**1. Add a config field to `Config`:**

```python
use_area_weighted_surface_loss: bool = False   # weight surface MSE by normalized point area
```

Add this immediately after `wallshear_z_weight` in the `Config` dataclass.

**2. Add CLI wiring** (in the argparse block):

```python
parser.add_argument("--use-area-weighted-surface-loss", action="store_true", default=False)
parser.add_argument("--no-use-area-weighted-surface-loss", dest="use_area_weighted_surface_loss", action="store_false")
```

**3. Thread `use_area_weighted_surface_loss` and the area tensor through `train_loss`.**

The `train_loss` function currently receives `batch` which has `batch.surface_x` — channel 6 of `surface_x` is area (in the loader concatenation: `[x, y, z, nx, ny, nz, area]`). Extract it:

```python
# Inside train_loss, after computing surface_loss_weight and surface_target:
if config.use_area_weighted_surface_loss:
    # Extract area from surface_x channel 6: [B, N, 1]
    area_raw = batch.surface_x[..., 6:7].clamp(min=0.0)
    # Normalize per case so weights sum to 1 over valid surface points
    mask_f = batch.surface_mask.unsqueeze(-1).float()  # [B, N, 1]
    area_sum = (area_raw * mask_f).sum(dim=1, keepdim=True).clamp(min=1e-8)  # [B, 1, 1]
    area_weights = area_raw / area_sum * mask_f  # [B, N, 1] — sums to ~1 per case
    area_weights = area_weights * batch.surface_mask.sum(dim=1, keepdim=True).unsqueeze(-1).float()
    # area_weights now is normalized so mean over valid pts ≈ 1.0
```

Then modify the loss computation. The `weighted_masked_mse_per_channel` function computes:
`(surface_diff_weighted * mask.unsqueeze(-1).float()).sum() / mask.float().sum().clamp_min(1) / 4.0`

You need to also multiply by `area_weights` before summing. Here is the modification to make in `weighted_masked_mse_per_channel` — add an optional `point_weights` parameter:

```python
def weighted_masked_mse_per_channel(
    pred: torch.Tensor,
    target: torch.Tensor,
    mask: torch.Tensor,
    channel_weights: tuple[float, ...],
    point_weights: torch.Tensor | None = None,   # [B, N, 1] normalized per-point weights
) -> tuple[torch.Tensor, list[float]]:
    """
    Returns the weighted scalar loss plus an unweighted per-channel mean for
    diagnostic logging. With all weights == 1 this is numerically equivalent
    to F.mse_loss(pred[mask], target[mask]).
    """
    weights = torch.tensor(list(channel_weights), device=pred.device, dtype=pred.dtype)
    diff = (pred - target) ** 2  # [B, N, C]
    diff_weighted = diff * weights.unsqueeze(0).unsqueeze(0)  # [B, N, C]
    mask_f = mask.unsqueeze(-1).float()  # [B, N, 1]
    if point_weights is not None:
        # point_weights: [B, N, 1], normalized so valid mean ≈ 1
        loss = (diff_weighted * mask_f * point_weights).sum() / max(mask.float().sum().item() * len(channel_weights), 1)
    else:
        loss = (diff_weighted * mask_f).sum() / mask.float().sum().clamp_min(1) / len(channel_weights)
    # Per-channel unweighted (for logging)
    per_ch = []
    for c in range(diff.shape[-1]):
        ch_val = (diff[..., c] * mask.float()).sum() / mask.float().sum().clamp_min(1)
        per_ch.append(float(ch_val.detach().cpu().item()))
    return loss, per_ch
```

Then pass `point_weights=area_weights if config.use_area_weighted_surface_loss else None` to the `weighted_masked_mse_per_channel` call inside `train_loss`.

Log `area_weights.max()` and `area_weights.min()` for the first batch each epoch for diagnostics.

**4. Run 2 arms** on 2 GPUs:

| GPU | Flag | `--wandb-name` |
|---|---|---|
| 0 | `--use-area-weighted-surface-loss` | `alphonse-area-wt-on` |
| 1 | (control — no flag) | `alphonse-area-wt-ctrl` |

```bash
cd target/
python train.py \
  --use-area-weighted-surface-loss \     # ARM 0 only
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group alphonse-area-weighted-loss \
  --wandb-name alphonse-area-wt-<SLUG>
```

**Beat target: abupt=12.74** (PR #66 thorfinn, W&B `gvigs86q`).

Watch especially: `test_primary/surface_pressure_rel_l2_pct` (target: 7.86→lower) and `test_primary/wall_shear_*` axes.

## Baseline (current yi best — PR #66 thorfinn, 6L/256d + W_y=2, W_z=2)

| Metric | Current best (`gvigs86q`) | AB-UPT |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **12.74** | — |
| `test_primary/surface_pressure_rel_l2_pct` | 7.86 | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | 12.86 | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | 13.14 | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | 11.29 | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | 15.15 | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | 15.05 | 3.63 |

**Reproduce (new base):**
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

## Results

Post results as a PR comment with:
1. All `test_primary/*_rel_l2_pct` and `full_val_primary/*` for both arms.
2. `area_weights` diagnostics (max/min area weight per case) — confirm the weights are non-trivial.
3. Per-epoch val trajectory for both arms.
4. W&B run IDs and URLs.

## Constraints
- Do **not** modify `data/*`, `pyproject.toml`, or `instructions/*`.
- Do **not** override `SENPAI_MAX_EPOCHS` or `SENPAI_TIMEOUT_MINUTES`.
- `test_primary/*` must **not** be NaN.

### CODE REVIEW

**Comment by morganmcg1** [comment]
**ADVISOR check-in** (2026-04-29)

W&B shows your runs have been active for ~3.3h. Here's the current snapshot:

| Run | W&B ID | State | abupt | sp | ws | vp |
|-----|--------|-------|-------|----|----|-----|
| alphonse-area-wt-on | bc58y54n | running | 89.95 | 77.23 | 92.22 | 74.55 |
| alphonse-area-wt-ctrl | wfcpx5m2 | running | 19.90 | 14.05 | 22.19 | 12.64 |
| alphonse-area-wt-on | d4a5yh71 | crashed | — | — | — | — |
| alphonse-area-wt-ctrl | i98ur8cq | crashed | — | — | — | — |

**Current baseline: 12.74** (thorfinn PR #66)

**Critical: The area-weighted-loss ON arm is diverging badly (89.95 vs ctrl 19.90).** The area weighting is destabilizing training — surface pressure error is 77.23% vs ctrl's 14.05%. All axes are 5-6x worse than control.

The initial crashed runs also suggest instability at startup.

Please report:
1. What area-weighting scheme are you using? (uniform weighting by face area? normalized? unnormalized?)
2. Are you using gradient clipping?
3. What is the loss scale / magnitude ratio between the area-weighted loss and the standard loss?

This does not look like a hyperparameter tuning issue — the area weighting is introducing a fundamentally unstable loss landscape. Consider: clamping area weights to [0.5, 2.0] to prevent extreme upweighting, or applying sqrt(area) instead of raw area.

**Comment by morganmcg1** [comment]
STUDENT alphonse:

## Mid-experiment update — naive (linear) area weighting diverged; relaunching tamed variant

**Timeline:** Launched both arms 14:59 UTC; control healthy; arm_on diverged ~step 11k (~epoch 2 boundary) and gradient norm later went to `Infinity`. Killed arm_on at 18:09 UTC.

**Cause:** With raw linear area weighting normalized to per-case mean=1, the dynamic range is severe — `area_weight_max ≈ 23.6` vs `area_weight_min ≈ 0.0024` (logged at step ~99). A few large patches dominate the surface gradient, and combined with `--wallshear-y/z-weight 2.0` the effective per-point gradient on those patches blew through `--clip-grad-norm 1.0` recovery. Both arms had a transient spike around the epoch-1→2 transition; ctrl recovered (loss 2.5→0.20→0.13), arm_on did not (loss stayed ~2.6, grad_norm went `Infinity`). val_primary/abupt of arm_on at val_1 = **89.95** (up from val_0=60.40); ctrl arm = **19.90** (down from val_0=35.20). W&B run: `bc58y54n` (arm_on, killed) and `wfcpx5m2` (ctrl, still running).

**Fix (relaunch on GPU 0 now):** Add two soft levers, both default-off so existing baseline is unchanged:
- `--area-weight-power` (default `1.0`): exponent applied to raw area before per-case normalization. `0.5` → sqrt-area weighting, compresses the max/mean ratio from ~23x to ~4.8x.
- `--area-weight-max-clamp` (default `0.0` = off): hard cap on the per-point weight, with re-normalization to keep mean ≈ 1.0 over valid points.

Tamed variant uses `--area-weight-power 0.5 --area-weight-max-clamp 5.0` — sqrt + cap. New W&B name: `alphonse-area-wt-on-tame`. This still tests the area-weighting hypothesis (large patches contribute more gradient) but bounds the dynamic range so training stays stable.

I've taken liberty here under the PR-instructions clause "you have liberty to modify the instructions to make them more specific and actionable if you think it will help". The naive arm result (diverged) will still be reported alongside the tamed arm so the divergence evidence is preserved. If you'd prefer a different recovery (e.g. lower LR, lower per-axis weights, different power), let me know and I'll swap in.

**Comment by morganmcg1** [comment]
## ADVISOR: Status Update — Alphonse Area-Weighted Loss

**Group:** alphonse-area-weighted-loss — Current State

| Run | abupt | sp | ws | vp | Status |
|-----|-------|----|----|-----|--------|
| alphonse-area-wt-on (`bc58y54n`) | 89.95 | 77.23 | 92.22 | 74.55 | CRASHED 3.3h |
| alphonse-area-wt-ctrl (`wfcpx5m2`) | 82.42 | 67.78 | 83.13 | 75.63 | FINISHED 4.6h |
| alphonse-area-wt-on (`d4a5yh71`) | N/A | — | — | — | crashed early |
| alphonse-area-wt-ctrl (`i98ur8cq`) | N/A | — | — | — | crashed early |
| alphonse-area-wt-on-tame (`v3acrl11`) | N/A | — | — | — | running 1.3h |

**Baseline: 12.74**

Situation is grim:
- Both area-wt-on arms crashed with very high abupt (89.95)
- The ctrl arm finished but at 82.42 — far above baseline
- Even the tame relaunch (`v3acrl11`) at 1.3h has no abupt yet

The ctrl arm finishing at 82.42 (vs baseline 12.74) is alarming — this suggests the entire experimental setup for this PR is misconfigured, not just the area-weighting. Both ctrl and treatment are failing badly. The original ctrl for the other PRs (thorfinn) runs fine, so something specific to this branch configuration is broken.

The area-weighted loss treatment may have a sign bug, normalisation issue, or incompatible loss scale. Even if the tame relaunch produces an abupt, all evidence points to a fundamentally broken setup. This experiment appears to be a negative result — recommend closing when the tame arm completes.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #95 because Closing: Area-weighted surface MSE loss did not improve performance. The 'tame' variant (v3acrl11, alphonse-tame-on) is at abupt=23.37 after 1.5h (sp=16.62, ws=26.72, vp=12.41) — significantly above baseline 12.74. Earlier attempts in this group also failed (area-wt-on at 89.95, area-wt-ctrl at 19.90 suggesting even the control was unstable). The physics-motivated area weighting introduces instability — larger faces receiving more weight appears to destabilize training rather than improve physical fidelity. The unweighted MSE is more stable for this geometry/resolution..
